pytorch_lightning/trainer/connectors/logger_connector/result.py

```diff
@@ -616,8 +616,8 @@ def cpu(self) -> "ResultCollection":
 
     def sync(self) -> None:
         for result_metric in self.result_metrics:
-            if result_metric.is_tensor:
-                result_metric.sync()
+            if result_metric.is_tensor and not result_metric._is_synced:
+                result_metric.sync(should_sync=not result_metric.meta.sync.rank_zero_only)
```

Comment:
n00b question: Is `result_metric` here a `ResultMetric` instance? I'm having trouble following this code.

Reply:
Yes.

A `ResultMetric` can wrap a Tensor or a `torchmetrics.Metric`. We only do this for tensors because `Metric`s should take care of syncing themselves.

Reply:
I see. After this PR, I'm seeing a few test failures on our end. The first is `ResultMetric does not have attribute `_is_synced`.

I worked around that error by defaulting `_is_synced` to `False` in the `ResultMetric` constructor, as per:
```
class ResultMetric(...):
  def __init__(self, ...):
    self._is_synced = False
    ...
```

But then the next error is similar, with `ResultMetric` does not have attribute `sync`. I ended up changing this to:
```
result_metric.meta.sync(value=result_metric.value)
```

which seems to pass our test cases. However, I'm not confident on whether these are the proper fixes to this issue, or if there's maybe something else I'm missing about how this code is supposed to work.


BTW, it's possible issues in this code were pre-existing. I believe the reason we're seeing this when merging this PR into Meta's codebase is because the `loops` state is now saved unconditionally, which appears to trigger this code path and leads to the above errors.

Reply:
Here's is a partial stack-trace of the first issue:
```
Traceback (most recent call last):
  File "/usr/local/fbcode/platform009/lib/python3.8/unittest/mock.py", line 1325, in patched
    return func(*newargs, **newkeywargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/stl/lightning/callbacks/tests/test_model_checkpoint.py", line 273, in test_ckpt_at_interval
    trainer.fit(model)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/trainer/trainer.py", line 743, in fit
    self._call_and_handle_interrupt(
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/trainer/trainer.py", line 694, in _call_and_handle_interrupt
    return trainer_fn(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/trainer/trainer.py", line 778, in _fit_impl
    results = self._run(model, ckpt_path=ckpt_path)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/trainer/trainer.py", line 1179, in _run
    results = self._run_stage()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/trainer/trainer.py", line 1268, in _run_stage
    return self._run_train()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/trainer/trainer.py", line 1300, in _run_train
    self.fit_loop.run()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/loops/base.py", line 191, in run
    self.advance(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/loops/fit_loop.py", line 231, in advance
    self.epoch_loop.run(data_fetcher)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/loops/base.py", line 191, in run
    self.advance(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/loops/epoch/training_epoch_loop.py", line 216, in advance
    self.trainer._call_callback_hooks("on_train_batch_end", batch_end_outputs, batch, batch_idx, **extra_kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/trainer/trainer.py", line 1576, in _call_callback_hooks
    fn(*args, **kwargs)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/trainer/callback_hook.py", line 183, in on_train_batch_end
    callback.on_train_batch_end(self, self.lightning_module, outputs, batch, batch_idx)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/callbacks/model_checkpoint.py", line 296, in on_train_batch_end
    self.save_checkpoint(trainer)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/callbacks/model_checkpoint.py", line 377, in save_checkpoint
    self._save_none_monitor_checkpoint(trainer, monitor_candidates)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/stl/lightning/callbacks/model_checkpoint.py", line 494, in _save_none_monitor_checkpoint
    super()._save_none_monitor_checkpoint(trainer, monitor_candidates)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/callbacks/model_checkpoint.py", line 663, in _save_none_monitor_checkpoint
    trainer.save_checkpoint(filepath, self.save_weights_only)
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/stl/lightning/callbacks/model_checkpoint.py", line 436, in _save_model
    trainer.checkpoint_connector.save_checkpoint(
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck
...
...
...
trainer/connectors/checkpoint_connector.py", line 485, in _get_loops_state_dict
    "fit_loop": self.trainer.fit_loop.state_dict(),
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/loops/base.py", line 284, in state_dict
    v.state_dict(destination, key + ".")
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/loops/base.py", line 284, in state_dict
    v.state_dict(destination, key + ".")
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/loops/base.py", line 287, in state_dict
    v.sync()
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/pytorch_lightning/trainer/connectors/logger_connector/result.py", line 619, in sync
    if result_metric.is_tensor and not result_metric._is_synced:
  File "/data/sandcastle/boxes/eden-trunk-hg-fbcode-fbsource/fbcode/buck-out/dev/gen/stl/lightning/callbacks/test_model_checkpoint#binary,link-tree/torch/nn/modules/module.py", line 1185, in __getattr__
    raise AttributeError("'{}' object has no attribute '{}'".format(
AttributeError: 'ResultMetric' object has no attribute '_is_synced'
```

Reply:
Yes. This was a bug in the existing implementation which had not been found before because we didn't save the `ResultCollection` state by default.

Have you checked your metrics version? Lightning requires

https://github.com/PyTorchLightning/pytorch-lightning/blob/29eb9cccf253201c8f83dc45d6c526267f8c9462/requirements.txt#L10

which is when `_is_synced` was introduced.

Reply:
Ah, good point. We're still on torchmetrics 0.4.0dev. Will update and report back!